### PR TITLE
Shadow Manifest

### PR DESCRIPTION
**Pull request by msftrubengu**
This PR introduces shadow manifests.

A shadow manifest is a file used to create a merged manifest that contains extra information about the package. For now, the shadow manifest contains icon information. The file is meant to be created programmatically and can be done via `ManifestShadow` from the `WinGetUtilnterop` nuget package.

The shadow will be added to the merged manifest when a directory contains a shadow manifest (ManifestType = shadow) and the create options allow using shadow manifest. If the flag is not set and a shadow manifest file is found it will fail. There can only be one shadow manifest file in the directory.

The shadow manifest gets processed and merged into the `Manifest` and the root YAML node. It will only merge information if the processed merged manifest doesn't already contain it. For example, if the manifest already contains a localization node with a specific `PackageLocale` that have `Icons` and the shadow manifest also contains a localization node with the same `PackageLocale` value with `Icons` it won't get merged.

A serialized shadow manifest looks like this (ignore values):
```
PackageIdentifier: Package.package
PackageVersion: 1.0
ManifestType: shadow
ManifestVersion: 1.5
PackageLocale: en-US
Icons:
- IconUrl: iconUrl
  IconFileType: fileType
  IconResolution: iconResolution
  IconTheme: iconTheme
  IconSha256: iconSha256
Localization:
- PackageLocale: es-MX
  Icons:
  - IconUrl: iconUrl-esMX
    IconFileType: fileType-esMX
    IconResolution: iconResolution-esMX
    IconTheme: iconTheme-esMX
    IconSha256: iconSha256-esMX
- PackageLocale: de-DE
  Icons:
  - IconUrl: iconUrl-de-DE
    IconFileType: fileType-de-DE
    IconResolution: iconResolution-de-DE
    IconTheme: iconTheme-de-DE
    IconSha256: iconSha256-de-DE
```
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/winget-cli/pull/4104)

### CODE REVIEW

**Comment by vedantmgoyal2009** [comment]
Is this related to, or a part of improvements to rest-source?

**Comment by msftrubengu** [comment]
@vedantmgoyal2009 no

File: src/AppInstallerCommonCore/Manifest/ManifestYamlPopulator.cpp [hunk 1]
```diff
@@ -1083,20 +1152,95 @@ namespace AppInstaller::Manifest
...
+            {
+                // Icons
+                if (!m_manifest.get().DefaultLocalization.Contains(Localization::Icons) &&
+                    shadowManifest.DefaultLocalization.Contains(Localization::Icons))
+                {
+                    m_manifest.get().DefaultLocalization.Add<Localization::Icons>(std::move(shadowManifest.DefaultLocalization.Get<Localization::Icons>()));
+
+                    YAML::Node key{ YAML::Node::Type::Scalar, "", YAML::Mark() };
+                    key.SetScalar("Icons");
+                    YAML::Node value = shadowNode["Icons"];
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/Manifest/ManifestYamlPopulator.cpp:1183]: We probably want case insensitivity here, but I think you might have to add that.

File: src/AppInstallerSharedLib/Public/winget/Yaml.h [hunk 2]
```diff
@@ -104,6 +104,10 @@ namespace AppInstaller::YAML
             return m_sequence->emplace_back(std::forward<Args>(args)...);
         }
 
+        // Merges sequence nodes. If both sequence have the specified key with the same value
+        // they will get merged together. All elements in sequence must have the key.
+        void MergeSequenceNode(Node& other, std::string_view key);
```
Comment (JohnMcPMS) [inline src/AppInstallerSharedLib/Public/winget/Yaml.h:109]: Same as other places; I think we probably want case-insensitivity.  It should probably be a parameter at this level.

File: src/AppInstallerSharedLib/Yaml.cpp [hunk 3]
```diff
@@ -398,19 +463,20 @@ namespace AppInstaller::YAML
...
 
-    void Node::MergeSequenceNode(Node& other, std::string_view key)
+    void Node::MergeSequenceNode(Node& other, std::string_view key, bool caseInsensitive)
     {
         Require(Type::Sequence);
         other.Require(Type::Sequence);
 
-        auto getKeyValue = [](const YAML::Node& node, std::string_view key) {
+        auto getKeyValue = [caseInsensitive](const YAML::Node& node, std::string_view key) {
                 auto keyNode = node[key];
```
Comment (JohnMcPMS) [inline src/AppInstallerSharedLib/Yaml.cpp:472]: Isn't this the part that needs to be case insensitive?

**Review by JohnMcPMS — APPROVED** [review]

**Pull request merged**